File: app/api/v1/endpoints/web/agents.py
"""
Follow these rules for all endpoints in this file:
1. Must return Pydantic models as JSON (no TemplateResponse or render()).
2. Must use FastAPI parameter types: Query, Path, Body, Depends, etc.
3. Must not parse inputs manually — let FastAPI validate and raise 422s.
4. Must use dependency-injected context for auth/user/project state.
5. Must not include database logic — delegate to a service layer (e.g. campaign_service).
6. Must not contain HTMX, Jinja, or fragment-rendering logic.
7. Must annotate live-update triggers with: # WS_TRIGGER: <event description>
"""

from typing import Annotated

from fastapi import (
    APIRouter,
    Body,
    Depends,
    HTTPException,
    Path,
    Query,
    status,
)
from sqlalchemy.ext.asyncio import AsyncSession
from starlette.datastructures import FormData

from app.core.authz import user_can
from app.core.deps import get_current_user, get_db
from app.core.exceptions import AgentNotFoundError
from app.core.services.agent_service import (
    get_agent_benchmark_summary_service,
    get_agent_by_id_service,
    get_agent_capabilities_service,
    get_agent_device_performance_timeseries,
    get_agent_error_log_service,
    list_agents_service,
    register_agent_full_service,
    test_presigned_url_service,
    toggle_agent_enabled_service,
    trigger_agent_benchmark_service,
    update_agent_config_service,
    update_agent_devices_service,
    update_agent_hardware_service,
)
from app.models.agent import OperatingSystemEnum
from app.models.user import User
from app.schemas.agent import (
    AdvancedAgentConfiguration,
    AgentBenchmarkSummaryOut,
    AgentCapabilitiesOut,
    AgentErrorLogOut,
    AgentListOut,
    AgentOut,
    AgentPerformanceSeriesOut,
    AgentPresignedUrlTestRequest,
    AgentPresignedUrlTestResponse,
    AgentRegisterModalContext,
    AgentToggleEnabledOut,
    AgentUpdateConfigOut,
    AgentUpdateDevicesOut,
    AgentUpdateHardwareOut,
    DevicePerformanceSeries,
)
from app.schemas.agent_error import AgentErrorOut

router = APIRouter(prefix="/agents", tags=["Agents"])


@router.get("", summary="List/filter agents")
async def list_agents(
    db: Annotated[AsyncSession, Depends(get_db)],
    search: Annotated[str | None, Query(description="Search by host name")] = None,
    state: Annotated[str | None, Query(description="Filter by agent state")] = None,
    page: Annotated[int, Query(ge=1, description="Page number")] = 1,
    size: Annotated[int, Query(ge=1, le=100, description="Page size")] = 20,
) -> AgentListOut:
    """Return a paginated, filterable list of agents."""
    agents, total = await list_agents_service(db, search, state, page, size)
    agents_out = [AgentOut.model_validate(a, from_attributes=True) for a in agents]
    return AgentListOut(
        agents=agents_out,
        page=page,
        size=size,
        total=total,
        total_pages=(total + size - 1) // size if total else 1,
        search=search,
        state=state,
    )


@router.get("/{agent_id}", summary="Agent detail")
async def agent_detail(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> AgentOut:
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return AgentOut.model_validate(agent, from_attributes=True)


@router.patch(
    "/{agent_id}",
    summary="Toggle agent enabled/disabled",
)
async def toggle_agent_enabled(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> AgentToggleEnabledOut:
    try:
        agent = await toggle_agent_enabled_service(agent_id, user, db)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e)) from e
    except AgentNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return AgentToggleEnabledOut(
        agent=AgentOut.model_validate(agent, from_attributes=True)
    )


@router.get(
    "/{agent_id}/benchmarks",
    summary="Agent benchmark summary",
)
async def agent_benchmark_summary(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> AgentBenchmarkSummaryOut:
    try:
        benchmarks_by_hash_type = await get_agent_benchmark_summary_service(
            agent_id, db
        )
    except AgentNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Agent not found"
        ) from e
    # Convert int keys to str for OpenAPI compatibility
    str_benchmarks = {str(k): v for k, v in benchmarks_by_hash_type.items()}
    return AgentBenchmarkSummaryOut(benchmarks_by_hash_type=str_benchmarks)


@router.post(
    "/{agent_id}/benchmark",
    summary="Trigger agent benchmark run (set to pending)",
)
# WS_TRIGGER: Notify agent benchmark status update
async def trigger_agent_benchmark(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> AgentOut:
    try:
        agent = await trigger_agent_benchmark_service(agent_id, user, db)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e)) from e
    except AgentNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return AgentOut.model_validate(agent, from_attributes=True)


@router.post(
    "/{agent_id}/test_presigned",
    summary="Validate presigned S3/MinIO URL for agent resource",
)
async def test_agent_presigned_url(
    agent_id: Annotated[int, Path(description="Agent ID")],
    payload: Annotated[AgentPresignedUrlTestRequest, Body(embed=True)],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> AgentPresignedUrlTestResponse:
    # Only admins can use this endpoint
    if (
        not user_can(user, "system", "create_users")
        and getattr(user, "role", None) != "admin"
    ):
        raise HTTPException(status_code=403, detail="Admin only")
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    valid = await test_presigned_url_service(str(payload.url))
    return AgentPresignedUrlTestResponse(valid=valid)


def _parse_agent_config_form(
    form: FormData,
) -> dict[str, int | str | None | bool | object]:
    data = {k: v for k, v in form.items() if not hasattr(v, "filename")}
    if "agent_update_interval" in data:
        try:
            data["agent_update_interval"] = str(int(str(data["agent_update_interval"])))
        except ValueError:
            data["agent_update_interval"] = "30"
    for key in ["use_native_hashcat", "enable_additional_hash_types"]:
        if key in data:
            v = str(data[key])
            data[key] = "true" if v == "true" else "false"
    for key in ["backend_device", "opencl_devices"]:
        if key in data and str(data[key]) == "":
            data[key] = ""
    return {
        "agent_update_interval": int(str(data.get("agent_update_interval", "30"))),
        "use_native_hashcat": str(data.get("use_native_hashcat", "false")) == "true",
        "backend_device": str(data.get("backend_device")) or None,
        "opencl_devices": str(data.get("opencl_devices")) or None,
        "enable_additional_hash_types": str(
            data.get("enable_additional_hash_types", "false")
        )
        == "true",
    }


@router.patch(
    "/{agent_id}/config",
    summary="Update agent advanced configuration",
)
async def update_agent_config(
    agent_id: Annotated[int, Path(description="Agent ID")],
    config: Annotated[AdvancedAgentConfiguration, Body(embed=True)],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> AgentUpdateConfigOut:
    try:
        await get_agent_by_id_service(agent_id, db)
    except AgentNotFoundError:
        raise HTTPException(status_code=404, detail="Agent not found") from None
    if not (
        getattr(user, "is_superuser", False)
        or user_can(user, "system", "update_agents")
    ):
        raise HTTPException(
            status_code=403, detail="Not authorized to update agent config"
        )
    try:
        updated_agent = await update_agent_config_service(agent_id, config, db)
    except AgentNotFoundError:
        raise HTTPException(status_code=404, detail="Agent not found") from None
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Update failed: {e}") from e
    return AgentUpdateConfigOut(
        agent=AgentOut.model_validate(updated_agent, from_attributes=True)
    )


@router.get("/{agent_id}/errors", summary="Agent error log")
async def agent_error_log(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> AgentErrorLogOut:
    errors = await get_agent_error_log_service(agent_id, db)
    errors_out = [AgentErrorOut.model_validate(e, from_attributes=True) for e in errors]
    return AgentErrorLogOut(errors=errors_out)


@router.patch(
    "/{agent_id}/devices",
    summary="Toggle enabled backend devices for agent",
)
async def toggle_agent_devices(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    enabled_indices: Annotated[
        list[int], Body(embed=True, description="List of enabled device indices")
    ],
) -> AgentUpdateDevicesOut:
    try:
        await update_agent_devices_service(agent_id, enabled_indices, user, db)
        agent = await get_agent_by_id_service(agent_id, db)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e)) from e
    except AgentNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return AgentUpdateDevicesOut(
        agent=AgentOut.model_validate(agent, from_attributes=True)
    )


@router.get(
    "/{agent_id}/performance",
    summary="Agent performance time series",
)
async def agent_performance(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> AgentPerformanceSeriesOut:
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    series: list[
        DevicePerformanceSeries
    ] = await get_agent_device_performance_timeseries(agent_id, db)
    return AgentPerformanceSeriesOut(series=series)


@router.post(
    "",
    summary="Register new agent and return token",
    description="Register a new agent from the web UI. Only admins can register agents. Returns agent and token.",
)
async def register_agent(
    host_name: Annotated[str, Body(..., description="Agent host name")],
    operating_system: Annotated[
        OperatingSystemEnum, Body(..., description="Operating system")
    ],
    client_signature: Annotated[str, Body(..., description="Client signature")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    custom_label: Annotated[str | None, Body(description="Custom label")] = None,
    devices: Annotated[
        str | None, Body(description="Comma-separated device list")
    ] = None,
    agent_update_interval: Annotated[
        int, Body(description="Agent update interval (seconds)")
    ] = 30,
    use_native_hashcat: Annotated[bool, Body(description="Use native hashcat")] = False,
    backend_device: Annotated[str | None, Body(description="Backend device")] = None,
    opencl_devices: Annotated[str | None, Body(description="OpenCL devices")] = None,
    enable_additional_hash_types: Annotated[
        bool, Body(description="Enable additional hash types")
    ] = False,
) -> AgentRegisterModalContext:
    if not (
        getattr(user, "is_superuser", False) or getattr(user, "role", None) == "admin"
    ):
        raise HTTPException(status_code=403, detail="Not authorized to register agents")
    agent_out, token = await register_agent_full_service(
        host_name=host_name,
        operating_system=operating_system,
        client_signature=client_signature,
        custom_label=custom_label,
        devices=devices,
        agent_update_interval=agent_update_interval,
        use_native_hashcat=use_native_hashcat,
        backend_device=backend_device,
        opencl_devices=opencl_devices,
        enable_additional_hash_types=enable_additional_hash_types,
        db=db,
    )
    return AgentRegisterModalContext(agent=agent_out, token=token)


@router.get("/{agent_id}/hardware", summary="Agent hardware detail")
async def agent_hardware(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> AgentOut:
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    resource = f"agent:{agent.id}"
    if not user_can(user, resource, "view_agent"):
        raise HTTPException(
            status_code=403, detail="Not authorized to view agent hardware"
        )
    return AgentOut.model_validate(agent, from_attributes=True)


async def _check_agent_update_permission(user: User, resource: str) -> None:
    if not user_can(user, resource, "update_agent") and not getattr(
        user, "is_superuser", False
    ):
        raise HTTPException(status_code=403, detail="Not authorized to update agent")


@router.patch(
    "/{agent_id}/hardware",
    summary="Update agent hardware limits and platform toggles",
    description="Update hardware-related advanced configuration fields for an agent. Only project admins or superusers may update.",
)
async def update_agent_hardware(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    hwmon_temp_abort: Annotated[
        int | None,
        Body(
            description="Temperature abort threshold in Celsius for hashcat (--hwmon-temp-abort)"
        ),
    ] = None,
    opencl_devices: Annotated[
        str | None,
        Body(
            description="The OpenCL device types to use for hashcat, separated by commas"
        ),
    ] = None,
    backend_ignore_cuda: Annotated[
        bool | None, Body(description="Ignore CUDA backend (--backend-ignore-cuda)")
    ] = None,
    backend_ignore_opencl: Annotated[
        bool | None, Body(description="Ignore OpenCL backend (--backend-ignore-opencl)")
    ] = None,
    backend_ignore_hip: Annotated[
        bool | None, Body(description="Ignore HIP backend (--backend-ignore-hip)")
    ] = None,
    backend_ignore_metal: Annotated[
        bool | None, Body(description="Ignore Metal backend (--backend-ignore-metal)")
    ] = None,
) -> AgentUpdateHardwareOut:
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    resource = f"agent:{agent.id}"
    await _check_agent_update_permission(user, resource)
    try:
        updated_agent = await update_agent_hardware_service(
            agent_id=agent_id,
            db=db,
            hwmon_temp_abort=hwmon_temp_abort,
            opencl_devices=opencl_devices,
            backend_ignore_cuda=backend_ignore_cuda,
            backend_ignore_opencl=backend_ignore_opencl,
            backend_ignore_hip=backend_ignore_hip,
            backend_ignore_metal=backend_ignore_metal,
        )
    except AgentNotFoundError as err:
        raise HTTPException(status_code=404, detail="Agent not found") from err
    except ValueError as err:
        raise HTTPException(status_code=400, detail=str(err)) from err
    return AgentUpdateHardwareOut(
        agent=AgentOut.model_validate(updated_agent, from_attributes=True)
    )


@router.get(
    "/{agent_id}/capabilities",
    summary="Show agent benchmark capabilities (table + graph)",
    responses={
        status.HTTP_200_OK: {
            "model": AgentCapabilitiesOut,
            "description": "Agent capabilities",
        },
        status.HTTP_404_NOT_FOUND: {
            "description": "Agent not found",
        },
    },
)
async def agent_capabilities(
    agent_id: Annotated[int, Path(description="Agent ID")],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> AgentCapabilitiesOut:
    agent = await get_agent_by_id_service(agent_id, db)
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Agent not found"
        )
    return await get_agent_capabilities_service(agent_id, db)

File: frontend/src/lib/schemas/agents.ts
/**
 * Agent schemas for CipherSwarm
 * Used by /api/v1/web/agents/* endpoints
 * Based on authoritative backend API schema
 */

import { z } from 'zod';
import { AgentState, AgentType, OperatingSystemEnum } from './base';

// Core agent schemas
/**
 * Agent output schema
 * Complete agent information including configuration and status
 */
export const AgentOut = z.object({
    id: z.number().int().describe('Agent ID'),
    host_name: z.string().describe('Agent hostname'),
    client_signature: z.string().describe('Client signature for identification'),
    custom_label: z.string().nullish().describe('Custom label for the agent'),
    token: z.string().describe('Agent authentication token'),
    state: AgentState.describe('Current agent state'),
    enabled: z.boolean().describe('Whether the agent is enabled'),
    advanced_configuration: z
        .union([z.record(z.string(), z.unknown()), z.null()])
        .nullish()
        .describe('Advanced configuration settings'),
    devices: z
        .union([z.array(z.string()), z.null()])
        .nullish()
        .describe('Available compute devices'),
    agent_type: AgentType.nullish().describe('Agent type'),
    operating_system: OperatingSystemEnum.describe('Operating system'),
    created_at: z.string().datetime().describe('Creation timestamp'),
    updated_at: z.string().datetime().describe('Last update timestamp'),
    last_seen_at: z
        .union([z.string().datetime(), z.null()])
        .nullish()
        .describe('Last seen timestamp'),
    last_ipaddress: z.string().nullish().describe('Last IP address'),
    projects: z.array(z.unknown()).default([]).describe('Projects associated with the agent'),
});
export type AgentOut = z.infer<typeof AgentOut>;

/**
 * Agent list output schema
 * Paginated list of agents with search and filtering
 */
export const AgentListOut = z.object({
    items: z.array(AgentOut).describe('List of agents'),
    total: z.number().int().describe('Total number of agents'),
    page: z.number().int().min(1).max(100).default(1).describe('Current page number'),
    page_size: z.number().int().min(1).max(100).default(20).describe('Number of items per page'),
    search: z.string().nullish().describe('Search query'),
    state: z.string().nullish().describe('Filter by agent state'),
});
export type AgentListOut = z.infer<typeof AgentListOut>;

/**
 * Agent dropdown item schema
 * Minimal agent information for dropdown selections
 */
export const AgentDropdownItem = z.object({
    id: z.number().int().describe('Agent ID'),
    display_name: z
        .string()
        .describe(
            'Agent display name, either custom_label or host_name if custom_label is not set'
        ),
    state: AgentState.describe('Agent state, either active, stopped, error, or offline'),
});
export type AgentDropdownItem = z.infer<typeof AgentDropdownItem>;

// Agent configuration schemas
/**
 * Advanced agent configuration schema
 * Detailed configuration options for agent behavior and performance
 */
export const AdvancedAgentConfiguration = z.object({
    agent_update_interval: z
        .number()
        .int()
        .nullish()
        .describe('The interval in seconds to check for agent updates'),
    use_native_hashcat: z
        .boolean()
        .nullish()
        .describe('Use the hashcat binary already installed on the client system'),
    backend_device: z
        .string()
        .nullish()
        .describe('The device to use for hashcat, separated by commas'),
    opencl_devices: z
        .string()
        .nullish()
        .describe('The OpenCL device types to use for hashcat, separated by commas'),
    enable_additional_hash_types: z
        .boolean()
        .describe('Causes hashcat to perform benchmark-all, rather than just benchmark'),
    hwmon_temp_abort: z
        .number()
        .int()
        .nullish()
        .describe('Temperature abort threshold in Celsius for hashcat (--hwmon-temp-abort)'),
    backend_ignore_cuda: z
        .boolean()
        .nullish()
        .describe('Ignore CUDA backend (--backend-ignore-cuda)'),
    backend_ignore_opencl: z
        .boolean()
        .nullish()
        .describe('Ignore OpenCL backend (--backend-ignore-opencl)'),
    backend_ignore_hip: z.boolean().nullish().describe('Ignore HIP backend (--backend-ignore-hip)'),
    backend_ignore_metal: z
        .boolean()
        .nullish()
        .describe('Ignore Metal backend (--backend-ignore-metal)'),
});
export type AdvancedAgentConfiguration = z.infer<typeof AdvancedAgentConfiguration>;

// Agent health and monitoring
/**
 * Agent health summary schema
 * Health status and metrics for the agent system
 */
export const AgentHealthSummary = z.object({
    total_agents: z.number().int().describe('Total number of agents'),
    online_agents: z.number().int().describe('Number of agents online (last seen <2min)'),
    total_campaigns: z.number().int().describe('Total number of campaigns'),
    total_tasks: z.number().int().describe('Total number of tasks'),
    total_hashlists: z.number().int().describe('Total number of hash lists'),
});
export type AgentHealthSummary = z.infer<typeof AgentHealthSummary>;

/**
 * Agent error log output schema
 * Error log entries for agent troubleshooting
 */
export const AgentErrorLogOut = z.object({
    errors: z.array(z.unknown()).describe('List of error entries'),
});
export type AgentErrorLogOut = z.infer<typeof AgentErrorLogOut>;

// Performance monitoring
/**
 * Device performance point schema
 * Single performance measurement point
 */
export const DevicePerformancePoint = z.object({
    timestamp: z.string().datetime().describe('Measurement timestamp'),
    value: z.number().describe('Performance value'),
});
export type DevicePerformancePoint = z.infer<typeof DevicePerformancePoint>;

/**
 * Device performance series schema
 * Time series performance data for a device
 */
export const DevicePerformanceSeries = z.object({
    device_name: z.string().describe('Device name'),
    data: z.array(DevicePerformancePoint).describe('Performance data points'),
});
export type DevicePerformanceSeries = z.infer<typeof DevicePerformanceSeries>;

/**
 * Agent performance series output schema
 * Performance time series data for an agent
 */
export const AgentPerformanceSeriesOut = z.object({
    series: z.array(DevicePerformanceSeries).describe('Performance series for each device'),
});
export type AgentPerformanceSeriesOut = z.infer<typeof AgentPerformanceSeriesOut>;

// Agent testing and validation
/**
 * Agent presigned URL test request schema
 * Request to test agent access to presigned URLs
 */
export const AgentPresignedUrlTestRequest = z.object({
    url: z.string().url().min(1).describe('The presigned S3/MinIO URL to test'),
});
export type AgentPresignedUrlTestRequest = z.infer<typeof AgentPresignedUrlTestRequest>;

/**
 * Agent presigned URL test response schema
 * Result of presigned URL test
 */
export const AgentPresignedUrlTestResponse = z.object({
    valid: z.boolean().describe('Whether the URL is valid and accessible'),
});
export type AgentPresignedUrlTestResponse = z.infer<typeof AgentPresignedUrlTestResponse>;

// Agent registration
/**
 * Agent register modal context schema
 * Context information for agent registration modal
 */
export const AgentRegisterModalContext = z.object({
    agent: AgentOut.describe('Registered agent information'),
    token: z.string().describe('Agent authentication token'),
});
export type AgentRegisterModalContext = z.infer<typeof AgentRegisterModalContext>;

// Agent update operations
/**
 * Agent toggle enabled output schema
 * Response for toggling agent enabled state
 */
export const AgentToggleEnabledOut = z.object({
    success: z.boolean().describe('Whether the operation was successful'),
    enabled: z.boolean().describe('New enabled state'),
});
export type AgentToggleEnabledOut = z.infer<typeof AgentToggleEnabledOut>;

/**
 * Agent update config output schema
 * Response for updating agent configuration
 */
export const AgentUpdateConfigOut = z.object({
    success: z.boolean().describe('Whether the operation was successful'),
    config: AdvancedAgentConfiguration.describe('Updated configuration'),
});
export type AgentUpdateConfigOut = z.infer<typeof AgentUpdateConfigOut>;

/**
 * Agent update devices output schema
 * Response for updating agent devices
 */
export const AgentUpdateDevicesOut = z.object({
    success: z.boolean().describe('Whether the operation was successful'),
    devices: z.array(z.string()).describe('Updated device list'),
});
export type AgentUpdateDevicesOut = z.infer<typeof AgentUpdateDevicesOut>;

/**
 * Agent update hardware output schema
 * Response for updating agent hardware information
 */
export const AgentUpdateHardwareOut = z.object({
    success: z.boolean().describe('Whether the operation was successful'),
    hardware: z.record(z.string(), z.unknown()).describe('Updated hardware information'),
});
export type AgentUpdateHardwareOut = z.infer<typeof AgentUpdateHardwareOut>;

// Request body schemas for API endpoints
/**
 * Register agent request body schema
 * Body for POST /api/v1/web/agents
 */
export const Body_register_agent_api_v1_web_agents_post = z.object({
    host_name: z.string().describe('Agent hostname'),
    custom_label: z.string().nullish().describe('Custom label for the agent'),
});
export type Body_register_agent_api_v1_web_agents_post = z.infer<
    typeof Body_register_agent_api_v1_web_agents_post
>;

/**
 * Test agent presigned URL request body schema
 * Body for POST /api/v1/web/agents/{agent_id}/test_presigned
 */
export const Body_test_agent_presigned_url_api_v1_web_agents__agent_id__test_presigned_post =
    AgentPresignedUrlTestRequest;
export type Body_test_agent_presigned_url_api_v1_web_agents__agent_id__test_presigned_post =
    z.infer<typeof Body_test_agent_presigned_url_api_v1_web_agents__agent_id__test_presigned_post>;

/**
 * Toggle agent devices request body schema
 * Body for PATCH /api/v1/web/agents/{agent_id}/devices
 */
export const Body_toggle_agent_devices_api_v1_web_agents__agent_id__devices_patch = z.object({
    devices: z.array(z.string()).describe('List of device identifiers'),
});
export type Body_toggle_agent_devices_api_v1_web_agents__agent_id__devices_patch = z.infer<
    typeof Body_toggle_agent_devices_api_v1_web_agents__agent_id__devices_patch
>;

/**
 * Update agent config request body schema
 * Body for PATCH /api/v1/web/agents/{agent_id}/config
 */
export const Body_update_agent_config_api_v1_web_agents__agent_id__config_patch = z.object({
    config: AdvancedAgentConfiguration.describe('Updated agent configuration'),
});
export type Body_update_agent_config_api_v1_web_agents__agent_id__config_patch = z.infer<
    typeof Body_update_agent_config_api_v1_web_agents__agent_id__config_patch
>;

/**
 * Update agent hardware request body schema
 * Body for PATCH /api/v1/web/agents/{agent_id}/hardware
 */
export const Body_update_agent_hardware_api_v1_web_agents__agent_id__hardware_patch = z.object({
    hardware: z.record(z.string(), z.unknown()).describe('Hardware information'),
});
export type Body_update_agent_hardware_api_v1_web_agents__agent_id__hardware_patch = z.infer<
    typeof Body_update_agent_hardware_api_v1_web_agents__agent_id__hardware_patch
>;
